design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+
+`uniqueItems` trait
+-------------------
+
+The [`uniqueItems` trait]
+
+> indicates that the items in a [`List`] MUST be unique.
+
+If the list shape is `sparse`, more than one `null` value violates this
+constraint.
+
+We will enforce this by copying _references_ to the `Vec`'s elements into a
+`HashSet` and checking that the sizes of both containers coincide. Smithy
+allows you to apply `uniqueItems` to _any_ list, [even if its members are not
+equipped with an equivalence relation], so we would have to manually check that
+the list's members are all [`Eq`].
+
+[`Eq`]: https://doc.rust-lang.org/std/cmp/trait.Eq.html
+[even if its members are not equipped with an equivalence relation]: https://github.com/awslabs/smithy/issues/1093
+[`List`]: https://awslabs.github.io/smithy/1.0/spec/core/model.html#list
+[`uniqueItems` trait]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#uniqueitems-trait
+[`is_match`]: https://docs.rs/regex/latest/regex/struct.Regex.html#method.is_match
+[`pattern` trait]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#pattern-trait
+[`aws_smithy_types::Blob`]: https://docs.rs/aws-smithy-types/latest/aws_smithy_types/struct.Blob.html
+[`aws_smithy_http_server::rejection::SmithyRejection`]: https://docs.rs/aws-smithy-http-server/latest/aws_smithy_http_server/rejection/enum.SmithyRejection.html
+[`aws_smithy_http_server::rejection::Deserialize`]: https://docs.rs/aws-smithy-http-server/latest/aws_smithy_http_server/rejection/struct.Deserialize.html
+[#1187]: https://github.com/awslabs/smithy-rs/issues/1187
+[`lazy_static!`](https://docs.rs/lazy_static/1.4.0/lazy_static/)
+
+Trait precedence and naming of the tuple struct
+-----------------------------------------------
+
+From [the spec]:
+
+> Some constraints can be applied to shapes as well as structure members. If a
+> constraint of the same type is applied to a structure member and the shape
+> that the member targets, the trait applied to the member takes precedence.
+
+```smithy
+structure ShoppingCart {
+    @range(min: 7, max:12)
+    numberOfItems: PositiveInteger
+}
+
+@range(min: 1)
+integer PositiveInteger
+```
+
+In the above example,
+
+> the range trait applied to numberOfItems takes precedence over the one
+> applied to PositiveInteger. The resolved minimum will be 7, and the maximum
+> 12.
+
+When the constraint trait is applied to a member shape, the tuple struct's name
+will be the PascalCased name of the member shape, `NumberOfItems`.
+
+[the spec]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html?highlight=enum#trait-precedence
+
+Unresolved questions
+--------------------
+
+1. Should we code-generate unsigned integer types (`u16`, `u32`, `u64`) when
+   the `range` trait is applied with `min` set to a value greater than or equal
+   to 0?
+    - A user has even suggested to use the `std::num::NonZeroUX` types (e.g.
```

Comment:
Personally, I'm all for using the `num` crate in situations like this.

Reply:
What specifically would you use from the `num` crate?